include/mvdtool/bits/mvd3_misc.hpp

```diff
@@ -164,16 +172,28 @@ inline std::vector<std::string> MVD3File::getMorphologies(const Range & range) c
 
 
 inline std::vector<std::string> MVD3File::getEtypes(const Range & range) const{
-    HighFive::DataSet index = _hdf5_file.getDataSet(did_cells_index_etypes);
-    HighFive::DataSet data = _hdf5_file.getDataSet(did_lib_data_etypes);
-    return resolve_index<std::string>(index, range, data);
+    if (!_tsv_file) {
+        HighFive::DataSet index = _hdf5_file.getDataSet(did_cells_index_etypes);
+        HighFive::DataSet data = _hdf5_file.getDataSet(did_lib_data_etypes);
+        return resolve_index<std::string>(index, range, data);
+    } else {
+        std::vector<std::string> morphologies = getMorphologies(range);
+        std::vector<std::string> me_combos = getMECombos(range);
+        return _tsv_file->getEtypes(me_combos, morphologies);
+    }
 }
 
 
-inline std::vector<std::string> MVD3File::getMtypes(const Range & range) const{
-    HighFive::DataSet index = _hdf5_file.getDataSet(did_cells_index_mtypes);
-    HighFive::DataSet data = _hdf5_file.getDataSet(did_lib_data_mtypes);
-    return resolve_index<std::string>(index, range, data);
+inline std::vector<std::string> MVD3File::getMtypes(const Range& range) const{
+    if (!_tsv_file) {
+        HighFive::DataSet index = _hdf5_file.getDataSet(did_cells_index_mtypes);
+        HighFive::DataSet data = _hdf5_file.getDataSet(did_lib_data_mtypes);
+        return resolve_index<std::string>(index, range, data);
+    } else {
+        std::vector<std::string> morphologies = getMorphologies(range);
```

Comment:
For the first point, I agree that there is some refactoring needed to avoid repeated code in all the readers.
For the second one, I wasn't aware of this issue. What you mean is to get the `morphologies` and `me_combos` one by one to use them to get the info from the `tsv` file?

Reply:
We can use mvd Ranges to get chunks at a time.
You can see something similar in the function `_atIndices()` in  `python/mvdtool.cpp`